src/test/kotlin/org/taktik/icure/services/external/rest/v1/controllers/core/DocumentControllerEndToEndTest.kt

```diff
@@ -0,0 +1,309 @@
+	}
+
+	"Updating main attachment with multipart should be allowed" {
+		val doc = createDocument(dataFactory.newDocumentNoAttachment())
+		val data = randomSmallAttachment()
+		val multipartBody =
+			MultipartBodyBuilder().apply {
+				asyncPart(
+					"attachment",
+					flowOf(DefaultDataBufferFactory.sharedInstance.wrap(data)).asPublisher(),
+					DataBuffer::class.java
+				).header(HttpHeaders.CONTENT_LENGTH, data.size.toString())
+					.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM_VALUE)
+			}.build()
+		val updated = client.put()
+			.uriWithVars(
+				"http://127.0.0.1:$port/$controllerRoot/${doc.id}/attachment/multipart",
+				mapOf(
+					"rev" to doc.rev,
+					"utis" to sampleUtis
+				)
+			)
+			.multipartContent()
+			.body(BodyInserters.fromMultipartData(multipartBody))
+			.retrieve()
+			.awaitDto()
+		updated.document.mainAttachment.shouldNotBeNull().apply {
+			shouldBeInCouch()
+			utis shouldBe sampleUtis
+		}
+		getMainAttachment(doc.id).toByteArray(true) shouldContainExactly data
+	}
+
+	"Attempting to modify a non-existing document should trigger creation of a new document" {
+		val dto = dataFactory.newDocumentNoAttachment()
+		val created = updateDocument(dto)
+		created.withoutDbUpdatedInfo shouldBe dto.withoutDbUpdatedInfo
+		val retrieved = getDocument(dto.id)
+		retrieved shouldBe created
+	}
+
+	"Changing a document attachment id in an update should trigger deletion of the attachment" {
+		val doc = createDocumentWithAttachment(dataFactory.newDocumentNoAttachment(), randomSmallAttachment(), null)
+		val updated = updateDocument(doc.copy(attachmentId = null)).document
+		updated.mainAttachment shouldBe null
+		ensureDeleted(doc.document, null)
+	}
+
+	"Changing a document attachment id in a bulk update should trigger deletion of the attachment" {
+		val docs = (1..10).map { i ->
+			createDocumentWithAttachment(dataFactory.newDocumentNoAttachment(i), randomSmallAttachment(), null)
+		}
+		val deleting = docs.take(5)
+		val updating = docs.drop(5)
+		val updatedById = bulkModify((deleting.map { it.copy(attachmentId = null) } + updating.map { it.changeNonAttachmentInfo() }).shuffled()).toList().associateBy { it.id }
+		deleting.forEach { original ->
+			updatedById[original.id].shouldNotBeNull().document.mainAttachment shouldBe null
+			ensureDeleted(original.document, null)
+		}
+		updating.forEach { original ->
+			updatedById[original.id].shouldNotBeNull().document.mainAttachment.shouldNotBeNull()
+		}
+	}
+
+	"Creating a document with initial main attachments uti should be allowed and they should be used when updating the attachment content" {
+		val initial = createDocument(dataFactory.newDocumentNoAttachment().copy(mainUti = xmlUti))
+		val withAttachment = updateMainAttachment(initial.id, initial.rev, randomBigAttachment(), null)
+		withAttachment.document.mainAttachment.shouldNotBeNull().let {
+			it.shouldBeInObjectStore()
+			it.utis shouldBe listOf(xmlUti)
+		}
+	}
```

Comment:
To ensure retro-compatibility, it would be interesting to add the following test cases as well : 
- `Getting an existing attachment which is too big for CouchDB starts an upload to Object Storage`
- `Getting an existing attachment which is not too big for CouchDB is a success`

Reply:
There is already a test to verify that loading a big attachment from couchdb through the attachment loader correctly triggers a migration task (upload to object storage and later update of the entity) in the [DataAttachmentLoaderTest](https://github.com/icure-io/icure-kraken-oss/blob/0d70402fb3f9b4d357ba12748430f61706e60f3f/src/test/kotlin/org/taktik/icure/asynclogic/objectstorage/DataAttachmentLoaderTest.kt#L87).

There are also tests to verify the loading of attachments works, regardless of where they are stored, in the [shared end-to-end tests](https://github.com/icure-io/icure-kraken-oss/blob/0d70402fb3f9b4d357ba12748430f61706e60f3f/src/test/kotlin/org/taktik/icure/services/external/rest/shared/controllers/core/DocumentControllerEndToEndTest.kt#L50).